Enable multi-select on gradio.Dropdown

dawoodkhan82 opened the pull request:
# Description
Add option to select multiple options on `gr.Dropdown`

Will be updating the colors to match other form components.

https://user-images.githubusercontent.com/12725292/210434026-737809ed-5e6c-4b10-97ff-e489bbfe72e4.mov




Please include: 
* relevant motivation
* a summary of the change 
* which issue is fixed. 
* any additional dependencies that are required for this change.

Closes: #1076

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

abidlabs commented on gradio/components.py:
```diff
@@ -1204,9 +1204,9 @@ def style(
+class Dropdown(Changeable, IOComponent, SimpleSerializable, FormComponent):
     """
-    Creates a dropdown of which only one entry can be selected.
+    Creates a dropdown of which only entries can be selected.
```
```suggestion
    Creates a dropdown of choices from which entries can be selected.
```

abidlabs replied:
Can we add a multiselect demo and link to it in the Demos line?

abidlabs commented:
Awesome, love the fact that the dropdown is typeable as well. Will be useful for very large number of choices. 

Some suggestions on the UI / UX side:

1. The appearance of `Dropdown(multiselect=False)` and `Dropdown(multiselect=True)` are very different. We should make them more consistent, with the only difference being that you can only choose a single element if `multiselect=False`. 

<img width="636" alt="image" src="https://user-images.githubusercontent.com/1778297/208828476-8b57e0e7-0071-4799-a3c8-0dbb24cc1a90.png">

<img width="624" alt="image" src="https://user-images.githubusercontent.com/1778297/208828523-929bd9a8-b053-46a5-8d47-983c48677709.png">

Among other things, the container with the `Dropdown` label is now completely gone. We should bring that back. But also, the entire design of the list of choices is different, the textbox part is different, etc.

2. You can't seem to currently use `Dropdown(multiselect=False)` as an output component the way you can with `Dropdown(multiselect=True)`.

Try running:

```py
import gradio as gr

with gr.Blocks() as demo:
    gr.Dropdown(["angola", "pakistan", "canada"], multiselect=True, value=["angola"])
    
demo.launch()
```

vs.

```py
import gradio as gr

with gr.Blocks() as demo:
    gr.Dropdown(["angola", "pakistan", "canada"], multiselect=False, value="angola")
    
demo.launch()
```

3. Typing doesn't filter choices. As I start typing "Pa", I would expect the choices to filter to just Pakistan, and that I could use the down arrow key and enter to go down and select it. However, this does not happen:
4. 
<img width="638" alt="image" src="https://user-images.githubusercontent.com/1778297/208828896-b24daf0f-2f64-4782-9583-3c6550dbd50d.png">

freddyaboulton commented on gradio/components.py:
```diff
@@ -1230,28 +1231,115 @@ def __init__(
+    def update(
+        value: Optional[Any] = _Keywords.NO_VALUE,
+        choices: Optional[str | List[str]] = None,
+        multiselect: bool = False,
```
Better to make `multiselect=None` otherwise any update to a multi-select dropdown will switch off the multiselect

aliabid94 replied:
We do not support `update` changing backend configuration yet, such as `multiselect`. `update` only allows changing frontend configuration. So please remove `multiselect` as an option to update

freddyaboulton commented on ui/packages/form/src/Dropdown.svelte:
```diff
@@ -1,27 +1,139 @@
 	export let disabled: boolean = false;
 	export let show_label: boolean;
+  
+    let input, 
```
There is a weird bug where hitting enter on the typeable field makes the whole app crash

![multiselect_enter](https://user-images.githubusercontent.com/41651716/209021520-a5802ee8-66ca-491f-8800-3c3492928ce7.gif)

freddyaboulton commented on ui/packages/form/src/Dropdown.svelte:
```diff
@@ -1,27 +1,139 @@
+    export let placeholder = '';
 	export let label: string;
-	export let value: string | undefined = undefined;
+	export let value: string | Array<string> | undefined = undefined;
```
Can't seem to update the value on a multi select drop down

![multiselect_update](https://user-images.githubusercontent.com/41651716/209021859-81c9bd16-8950-4788-8e07-7c356a3d201e.gif)


This is the code I ran

```python

import gradio as gr
import random

with gr.Blocks() as demo:
    dd = gr.Dropdown(["angola", "pakistan", "canada"], multiselect=True, value=["angola"], interactive=True)
    update = gr.Button(value="Update")
    update.click(lambda: gr.Dropdown.update(
                                            value=[random.choice(["angola", "pakistan", "canada"])],
                                            multiselect=True),
                 inputs=None,
                 outputs=[dd])

demo.launch()
```

freddyaboulton reviewed (commented):
@dawoodkhan82 This is so cool! It's great that developers can update whether or not multi-select is enabled via `gr.Dropdown.update` as well.

Left comments on some bugs I think I found.

In addition, I agree with @abidlabs comment about filtering with what's been typed and the look being too different from the existing dropdown.

freddyaboulton commented on ui/packages/atoms/src/Block.svelte:
```diff
@@ -45,7 +45,7 @@
 	data-testid={test_id}
 	id={elem_id}
 	class:!hidden={visible === false}
-	class="gr-block gr-box relative w-full overflow-hidden {styles[
```
How come you made this change?

abidlabs commented:
Was just testing it out @dawoodkhan82, UI looks nice! Just wanted to point out that if you click on the scrollbar, an entry named "undefined" shows up in the dropdown:

<img width="632" alt="image" src="https://user-images.githubusercontent.com/1778297/209739317-75890b4a-59f5-4587-a473-ee71adcdfc78.png">

Also maybe it's just me, but I think having just 2 entries visible at a time in the Dropdown select is too little. It should be at least 5-6 imo

dawoodkhan82 commented:
> Was just testing it out @dawoodkhan82, UI looks nice! Just wanted to point out that if you click on the scrollbar, an entry named "undefined" shows up in the dropdown:
> 
fixed. good catch, weird edge case...
> <img alt="image" width="632" src="https://user-images.githubusercontent.com/1778297/209739317-75890b4a-59f5-4587-a473-ee71adcdfc78.png">
> 
> Also maybe it's just me, but I think having just 2 entries visible at a time in the Dropdown select is too little. It should be at least 5-6 imo
> 
updated max height
> Is the typeability gone?

Typeability is something I had enabled by accident. I can look into making it work. Is it something we definitely want to add?

abidlabs commented:
> Typeability is something I had enabled by accident. I can look into making it work. Is it something we definitely want to add?

Yeah I'm imagining the case with 100s of choices (this has been asked about frequently). It's going to be hard to choose the right choices without typeability I believe

abidlabs commented:
Super nice @dawoodkhan82!! I tested it out and multiselect dropdown works really well. Just a couple of minor points I noticed:

(1) The blue colors used as the  background for the selected choices feels a little "different" than our usual palette. How about some faint orange colros?

<img width="868" alt="image" src="https://user-images.githubusercontent.com/1778297/210009821-26aa8f2d-45af-4ad4-96fd-344d787249c0.png">

(2) In dark mode, the background color of the dropdown choices is still white: 

<img width="963" alt="image" src="https://user-images.githubusercontent.com/1778297/210009930-b8fd95d5-6d3d-4c12-9fe4-92cea35b0556.png">

We should make it the same color as what shows if `multiselect=False`

<img width="473" alt="image" src="https://user-images.githubusercontent.com/1778297/210009945-897ea88a-30e0-49ff-bf4c-e4577a27ecc9.png">

(3) The `multiselect` mode doesn't respect `interactive=False` in the way that I expect

If `multiselect=False`, the Dropdown choice is displayed, and the dropdown itself cannot be clicked on:

<img width="628" alt="image" src="https://user-images.githubusercontent.com/1778297/210010025-d08a17d2-7d08-4f2f-90ab-dadbeea885b6.png">

If `multiselect=True`, the Dropdown choice are displayed, but the dropdown can still be clicked on:

<img width="658" alt="image" src="https://user-images.githubusercontent.com/1778297/210010111-205e9c0c-2104-4e1a-bd7f-ccf22396a8cd.png">

We should make it un-changeable so it is truly a non-interactive component

Other than these, LGTM! Would be good to get a review on the frontend code, perhaps by @aliabid94?

dawoodkhan82 commented:
@abidlabs fixed all 3

aliabid94 commented:
great work. styling is still a bit off, I'll just make changes directly to branch.

aliabid94 commented:
Just looking at this, I feel like should be a completely different element than Dropdown, e.g. a `gr.Select`. It has a different type (List of strings vs string), different purpose, and different visual layout than Dropdown. There's so little in common that all the logic branches whether or not it is a multiselect. @abidlabs @dawoodkhan82

abidlabs commented:
> Just looking at this, I feel like should be a completely different element than Dropdown, e.g. a gr.Select. It has a different type (List of strings vs string), different purpose, and different visual layout than Dropdown. There's so little in common that all the logic branches whether or not it is a multiselect. @abidlabs @dawoodkhan82

From a user's perspective, it feels "natural" to me to have a multiselect variant of the dropdown alongside a regular single-select dropdown. But I don't have a strong opinion, I'm fine with a separate component. If we do end up making it a different component, I think `gr.Multiselect` is more of an appropriate name

dawoodkhan82 commented:
Synced with @aliabid94 , there's changes he suggested which makes sense and I'll implement:
- [x] change the styling of the selected option bubbles to match the radio button
- [ ] separate the front end code into a separate "Multiselect.svelte" component
- [ ] re-implement the multiselect dropdown to use `datalist`
- [ ] Maybe separate the component completely (frontend + backend)? Not sure about this, I think for the same reasons as @abidlabs mentioned, it makes sense from a developer's pov to have it be the same component.

abidlabs commented:
This is going to be a really slick component!

dawoodkhan82 commented:
synced with @aliabid94 and @pngwn. Decided to move forward with the custom drop-down. Will look into combining the custom dropdown with datalist to improve accessibility, along with the other forms.

aliabid94 commented on ui/packages/app/src/components/Column/Column.svelte:
```diff
@@ -12,7 +12,7 @@
 <div
 	id={elem_id}
-	class="overflow-hidden flex flex-col relative col {create_classes(style)}"
+	class="flex flex-col relative col {create_classes(style)}"
```
I can't think of any harm that's done by removing `overflow-hidden` here, but can you confirm @pngwn?

aliabid94 replied:
This is done to allow the dropdown to extend past the parent.

pngwn replied:
We just need to make sure that this wasn't responsible for enforcing rounded borders (which I don't think it is/was, Block does that). Otherwise safe to remove.

aliabid94 commented on ui/packages/form/src/MultiSelect.svelte:
```diff
@@ -0,0 +1,195 @@
+
+	function handleKeyup(e: any) {
+		// enter key
+		if (e.keyCode === 13) {
```
keycode is deprecated in favor of `event.key === 'Enter'`

aliabid94 commented on ui/packages/form/src/MultiSelect.svelte:
```diff
@@ -0,0 +1,195 @@
+				inputValue = "";
+			}
+		}
+		if ([38, 40].includes(e.keyCode)) {
```
same here

abidlabs commented on CHANGELOG.md:
```diff
@@ -20,6 +20,14 @@ demo = gr.Interface(reverse, gr.Text(), gr.Text())
 * Added `title` argument to `TabbedInterface` by @MohamedAliRashad in [#2888](https://github.com/gradio-app/gradio/pull/2888)
+* Added `multiselect` option to `Dropdown` by @dawoodkhan82 in [#2871](https://github.com/gradio-app/gradio/pull/2871)
+
+### With `multiselect` set to `true` a user can now select multiple options from the dropdown menu.
```
```suggestion
### With `multiselect` set to `true` a user can now select multiple options from the `gr.Dropdown` component.
```

abidlabs reviewed (approved):
Looks beautiful! Nit: can we add backend tests to ensure that preprocessing/postprocess multiselect works as we expect it to?
